cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/ExternalDatasets.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.common;
+
+import co.cask.cdap.api.Admin;
+import co.cask.cdap.api.data.batch.Input;
+import co.cask.cdap.api.data.batch.InputFormatProvider;
+import co.cask.cdap.api.data.batch.Output;
+import co.cask.cdap.api.data.batch.OutputFormatProvider;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import com.google.common.base.Throwables;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class is used to create external datasets to track external sources/sinks.
+ */
+public final class ExternalDatasets {
+
+  /**
+   * If the input is an external source then an external dataset is created for tracking purpose and returned.
+   * If the input is a regular dataset or a stream then it is already trackable, hence same input is returned.
+   *
+   * @param admin {@link Admin} used to create external dataset
+   * @param input input to be tracked
+   * @return an external dataset if input is an external source, otherwise the same input that is passed-in is returned
+   */
+  public static Input makeTrackable(Admin admin, Input input) {
+    // If input is not an external source, return the same input as it can be tracked by itself.
+    if (!(input instanceof Input.InputFormatProviderInput)) {
+      return input;
+    }
+
+    // Input is an external source, create an external dataset so that it can be tracked.
+    String inputName = input.getName();
+    InputFormatProvider inputFormatProvider = ((Input.InputFormatProviderInput) input).getInputFormatProvider();
+    Map<String, String> inputFormatConfiguration = inputFormatProvider.getInputFormatConfiguration();
+
+    // Input is a dataset that implements input format provider,
+    // this too can be tracked by itself without creating an external dataset
+    if (inputFormatProvider instanceof Dataset) {
+      return input;
+    }
+
+    try {
+      // Create an external dataset for the input format for lineage tracking
+      String datasetName = "ext-" + inputName;
+      Map<String, String> arguments = new HashMap<>();
+      arguments.put("input.format.class", inputFormatProvider.getInputFormatClassName());
+      arguments.putAll(inputFormatConfiguration);
+      if (!admin.datasetExists(datasetName)) {
```

Comment:
With this, is is possible to have one pipeline writing to an external data store as a Sink, another reading from the same external store as it Source, and have them represented by the same external dataset? For example, pipeline A writing to a Kafka topic, pipeline B reading from that topic, and the alias would be the topic name, say myTopic, so that they become the same external dataset?

It seems that in this case, you would create an externalDataset ext-myTopic that - assuming A is created first, has the required output arguments for the sink, but then when B is deployed, the external dataset already exists and you will not record the input arguments. So even though ext-myTopic implements InputFormatProvider, it will fail to work as one. 

But we do want to support that use case, right?